chapter_explainability/global_model_analysis.md

```diff
@@ -0,0 +1,265 @@
+    
+    def predict(self, data, shuffle_x=None):
+        Y_hat, Y, X = [], [], []
+        for x, y in data:
+            if shuffle_x is not None:
+                x_col = x[:,shuffle_x]
+                x[:, shuffle_x] = x_col[torch.randperm(x.shape[0])]
+            Y_hat.append(self(x))
+            Y.append(y)
+            X.append(x)
+        acc = self.accuracy(torch.cat(Y_hat, 0), torch.cat(Y, 0)).numpy()
+        return acc, torch.cat(Y_hat, 0), torch.cat(Y, 0), torch.cat(X, 0)
+```
+
+Let's train and validate this model.
+
+```{.python .input  n=2}
+model = MlpLoan(num_outputs=2, lr=0.001, wd=1e-5)
+trainer = d2l.Trainer(max_epochs=20)
+trainer.fit(model, data)
+print("Validation accuracy is:", model.predict(data.val_dataloader())[0])
+```
+
+As can be seen, our model obtains a decent accuracy score on the validation set. Yet, instead of stopping at this point, we can further scrutinize our model by asking: (1) What type of errors have been made by the classification model? (2) How does the model perform on different groups of loan borrowers? (3) How does each feature contribute to the model performance?
+
+## Error Analysis
+As what we have done, aggregate metrics such as accuracy and mean squared error on the entire dataset are usually adopted to assess the effectiveness of machine learning models. These metrics help us understand the overall model performances but do not reflect the specifics of the errors. In particular, the model accuracy may be non-uniform across subgroups of data, and there might exist cases where the model fails more frequently. Pinpointing such cases via error analysis is helpful in model debugging and the detection of root causes of errors. 
+
+In the following, we will demonstrate how to identify the error distributions across different cohorts via error analysis.  We define a *GlobalExplainer* class which requires to be fed with a black-box model and a dataset, and save it for later use.
+
+```{.python .input  n=5}
+#@save
+class GlobalExplainer: 
+    def __init__(self, black_box_model, data):
+        self.black_box_model = black_box_model
+        self.data = data
+        
+explainer = GlobalExplainer(model, data)
+```
+
+The function below is used to record the errors and features of interest.
+
+```{.python .input  n=3}
+@d2l.add_to_class(GlobalExplainer)
+def error_analysis(self, feat_of_interest=None):
+    acc, pred, Y, X = self.black_box_model.predict(self.data.val_dataloader())
+    error = pd.DataFrame()
+    pred = torch.argmax(pred, -1)
+    error['Actual_Label'] = ["Fully-Paid" if i==1 else "Charge-Off" 
+                                    for i in Y.cpu().detach().numpy()]
+    error['Predicted_Label'] = ["Fully-Paid" if i==1 else "Charge-Off" 
+                                  for i in pred.cpu().detach().numpy()]
+    error['Status'] = ["Misclassified" if i==False else "Correctly_Classified" 
+                                for i in (pred == Y).cpu().detach().numpy()]
```

Comment:
Is it necessary to invoke cpu()? It looks to me the data & model are already using CPU?